vnext/ReactUWP/Modules/DeviceInfoModule.cpp

```diff
@@ -14,42 +12,83 @@ namespace uwp {
 // DeviceInfo
 //
 
-void DeviceInfo::update() {
-  auto displayInfo = winrt::Windows::Graphics::Display::DisplayInformation::
-      GetForCurrentView();
-  winrt::Windows::UI::ViewManagement::UISettings uiSettings;
+DeviceInfo::DeviceInfo() {
+  m_dimensions = getDimensions(nullptr, nullptr);
+}
 
-  auto const &window = winrt::Windows::UI::Xaml::Window::Current().CoreWindow();
+folly::dynamic DeviceInfo::getDimensions(
+    winrt::Windows::Graphics::Display::DisplayInformation displayInfo,
+    winrt::Windows::UI::Core::CoreWindow window) {
+  winrt::Windows::UI::ViewManagement::UISettings uiSettings;
 
   // TODO: get parent element (not window) size
-  m_dimensions = folly::dynamic::object(
+  return folly::dynamic::object(
       "windowPhysicalPixels",
-      folly::dynamic::object("width", window.Bounds().Width)(
-          "height", window.Bounds().Height)(
-          "scale", static_cast<int>(displayInfo.ResolutionScale()) / 100)(
-          "fontScale", uiSettings.TextScaleFactor())(
-          "densityDpi", displayInfo.LogicalDpi()))(
+      folly::dynamic::object(
+          "width", window != nullptr ? window.Bounds().Width : 100)(
+          "height", window != nullptr ? window.Bounds().Height : 100)(
+          "scale",
+          displayInfo != nullptr
+              ? static_cast<int>(displayInfo.ResolutionScale()) / 100
+              : 1)("fontScale", uiSettings.TextScaleFactor())(
+          "densityDpi",
+          displayInfo != nullptr ? displayInfo.LogicalDpi() : 100))(
       "screenPhysicalPixels",
-      folly::dynamic::object("width", displayInfo.ScreenWidthInRawPixels())(
-          "height", displayInfo.ScreenHeightInRawPixels())(
-          "scale", static_cast<int>(displayInfo.ResolutionScale()) / 100)(
-          "fontScale", uiSettings.TextScaleFactor())(
-          "densityDpi", displayInfo.LogicalDpi()));
+      folly::dynamic::object(
+          "width",
+          displayInfo != nullptr ? displayInfo.ScreenWidthInRawPixels() : 100)(
+          "height",
+          displayInfo != nullptr ? displayInfo.ScreenHeightInRawPixels() : 100)(
+          "scale",
+          displayInfo != nullptr
+              ? static_cast<int>(displayInfo.ResolutionScale()) / 100
+              : 1)("fontScale", uiSettings.TextScaleFactor())(
+          "densityDpi",
+          displayInfo != nullptr ? displayInfo.LogicalDpi() : 100));
+}
+
+void DeviceInfo::update() {
+  m_dimensions = getDimensions(
+      winrt::Windows::Graphics::Display::DisplayInformation::
+          GetForCurrentView(),
+      winrt::Windows::UI::Xaml::Window::Current().CoreWindow());
 }
 
 //
 // DeviceInfoModule
 //
 const char *DeviceInfoModule::name = "DeviceInfo";
+DeviceInfoModule *DeviceInfoModule::s_currentInstance = nullptr;
 
 DeviceInfoModule::DeviceInfoModule(std::shared_ptr<DeviceInfo> deviceInfo)
-    : m_deviceInfo(std::move(deviceInfo)) {}
+    : m_deviceInfo(std::move(deviceInfo)) {
+  s_currentInstance = this;
```

Comment:
Do you need to clear this in the destructor?

Or instead of using a global can it detect not on a ui thread and use some fallback data?

Reply:
my assumption was that because it's a raw pointer it wouldn't take a reference, but maybe that's not correct.

It's possible to detect that we are not on the UI thread, but the challenge with that is more about knowing when to get the real data.

One thought I had was to have an interface that can be implemented by modules that need to be notified when the UI thread becomes available. But I wanted to do the easy thing first to get input on the approach.

Reply:
I see now that your point was more along correctness of this not a circular reference. the lifetime of DeviceInfoModule isn't necessarily the same as the lifetime of the app and so the pointer could reference a dead object.

I'll look into alternatives for this.

Reply:
In office code we theoretically have multiple instances / root views / uithreads that come and go from the process, so that one ever existed doesn't mean it continues to.